### PR TITLE
feat(sql): add ntile, cume_dist, and nth_value window functions

### PR DESCRIPTION
**Pull request by jovfer**
Addresses https://github.com/questdb/roadmap/issues/120 (Gap 2).
This PR covers only Gap 2 and does not close the roadmap issue.

## Summary

- Add `ntile(n)` window function — distributes rows of an ordered partition into `n` approximately equal buckets and returns the 1-based bucket number.
- Add `cume_dist()` window function — returns the cumulative distribution:
  rows at or before the current row (including peers) divided by total rows in the partition.
- Add `nth_value(expr, n)` window function — returns the `n`-th value (1-based) within the current window frame, or NULL when `n` exceeds the frame size. The argument must be `DOUBLE` in this PR; `LONG` and `TIMESTAMP` overloads are deferred to a follow-up that will mirror the existing per-type factory split used by `first_value` / `last_value` (`*Double`, `*Long`, `*Timestamp`).
- All three support `PARTITION BY` and `ORDER BY`.
- `nth_value` supports `ROWS` and `RANGE` frames (bounded and unbounded).

## Side effect

`SqlCodeGenerator` now forwards the `ORDER BY` expression list to `WindowFunction.initRecordComparator` on the grouped-comparator path. This was previously `null`, which left `percent_rank()`'s EXPLAIN output incomplete on non-timestamp `ORDER BY`. The new functions need the list for their plan rendering, and threading it through also tightens the existing `percent_rank` plan output.

## Tradeoffs / limitations

- `ntile` and `cume_dist` run in two passes and maintain per-partition state in a `Map`; memory scales with partition cardinality.
- `cume_dist` partitioned: per-partition deferred-offset slices live in a single shared native buffer and are not released until the cursor closes. Worst case is O(total rows) of native memory when peer groups span entire partitions (e.g. `ORDER BY` a constant column).
- `ntile` and `cume_dist` reject explicit `ROWS` / `RANGE` / `GROUPS` frame clauses — they are always partition-scoped.
- `nth_value` accepts only `DOUBLE` first argument in this PR; `LONG` and `TIMESTAMP` overloads will follow.
- `nth_value` requires `n` to be a compile-time constant; a non-constant expression for `n` is rejected.
- `nth_value` rejects `IGNORE NULLS` / `RESPECT NULLS` and does not support `FROM FIRST` / `FROM LAST`.
- `nth_value` `RANGE ... CURRENT ROW` follows the project-wide QuestDB convention shared with `sum` / `avg` / `min` / `max` / `first_value` / `last_value`: a frame ending at `CURRENT ROW` does not look ahead to peer rows that share the same `ORDER BY` value. This diverges from the SQL standard / PostgreSQL on tied `ORDER BY` values; revisiting peer semantics is tracked as a project-wide follow-up that should cover all `RANGE`-supporting window factories together.

## Related PRs

- #7037 — adds `nth_value(<long>, n)` and `nth_value(<timestamp>, n)` overloads on top of this branch.
- https://github.com/questdb/documentation/pull/430

## Test plan

- Explicit expected-value tests for each function
- Partition and non-partition variants
- Edge cases: single row, n > row count, all peers (cume_dist), n=1 (ntile, nth_value)
- Error cases: invalid arguments, framing rejection (ntile/cume_dist), IGNORE NULLS rejection (nth_value), frame-bound overflow (`|rowsLo|` / `|rowsHi|` > `Integer.MAX_VALUE`)
- NULL data handling, including NULL propagation through all five partitioned nth_value variants
- RANGE frame tests for nth_value, including designated-timestamp rejection
- Non-window context error path tested for all factories
- Included in WINDOW_ONLY_FUNCTIONS parametric test suite

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by mtopolnik** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 865 / 1004 (86.16%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/NthValueDoubleWindowFunctionFactory.java|562|676|83.14%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/NtileFunctionFactory.java|104|113|92.04%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/CumeDistFunctionFactory.java|199|215|92.56%|

**Comment by mtopolnik** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 934 / 1069 (87.37%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/NthValueDoubleWindowFunctionFactory.java|613|717|85.50%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/CumeDistFunctionFactory.java|218|241|90.46%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/NtileFunctionFactory.java|103|111|92.79%|

**Comment by mtopolnik** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 972 / 1118 (86.94%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/NthValueDoubleWindowFunctionFactory.java|646|755|85.56%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/CumeDistFunctionFactory.java|221|247|89.47%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/NtileFunctionFactory.java|105|116|90.52%|

File: core/src/main/java/io/questdb/griffin/engine/functions/window/NthValueDoubleWindowFunctionFactory.java [hunk 1]
```diff
@@ -0,0 +1,1668 @@
...
+            } else {
+                sink.val("unbounded preceding");
+            }
+            sink.val(" and ");
+            if (minDiff == 0) {
+                sink.val("current row");
+            } else {
+                sink.val(minDiff).val(" preceding");
+            }
+            sink.val(')');
```
[the remark this answers is not in the document]
Comment (jovfer) [reply core/src/main/java/io/questdb/griffin/engine/functions/window/NthValueDoubleWindowFunctionFactory.java:741]: @coderabbitai it's default pattern for now, see AvgDoubleWindowFunctionFactory or any other function in master

**Review by bluestreak01 — APPROVED** [review]

**Comment by mtopolnik** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 1106 / 1192 (92.79%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|25|29|86.21%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/CumeDistFunctionFactory.java|235|261|90.04%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/NtileFunctionFactory.java|125|136|91.91%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/NthValueDoubleWindowFunctionFactory.java|721|766|94.13%|
